km/load_elf.c

```diff
@@ -263,3 +240,97 @@ uint64_t km_load_elf(km_elf_t* e)
    km_guest.km_load_adjust = adjust;
    return adjust;
 }
+
+/*
+ * Open ELF file for KM.
+ * Note: This is a place where we need to really validate what we are looking
+ * at since. Potentially a bad actor could give us a bad elf file for the guest
+ * in an attempt to corrupt KM. We need to prevent that.
+ * For now there are bsic tests. 
+ */
+km_elf_t*
+km_open_elf_file(const char *path)
+{
+   // Open ELF file
+   km_elf_t *elf = (km_elf_t *) malloc(sizeof(km_elf_t));
+   memset(elf, 0, sizeof(km_elf_t));
+   elf->path = path;
+   if ((elf->file = fopen(elf->path, "r")) == NULL) {
+      km_err(2, "Cannot open %s", path);
+   }
+
+   // Read ELF header
+   elf->ehdr = (Elf64_Ehdr *) malloc(sizeof(Elf64_Ehdr));
+   int nread = fread(elf->ehdr, 1, sizeof(Elf64_Ehdr), elf->file);
+   if (nread < sizeof(Elf64_Ehdr)) {
+      km_err(2, "Cannot read EHDR %s", path);
+   }
+
+   // Validate Elf Header
+   unsigned char* wp = &elf->ehdr->e_ident[0];
+   if (memcmp(wp, ELFMAG, SELFMAG) != 0) {
+      km_errx(2, "EHDR magic number mismatch %s", path);
+   }
+   if (elf->ehdr->e_ident[EI_CLASS] != ELFCLASS64) {
+      km_errx(2, "Not a 64 bit ELF %s", path);
+   }
+   if (elf->ehdr->e_version != EV_CURRENT) {
+      km_errx(2, "Not current ELF version %s", path);
+   }
+
+   elf->phdr = malloc(elf->ehdr->e_phentsize * elf->ehdr->e_phnum);
+   if (fseek(elf->file, elf->ehdr->e_phoff, SEEK_SET) != 0) {
+      km_err(2, "Cannot seek to PHDR %s", path);
+   }
+   nread = fread(elf->phdr, 1, elf->ehdr->e_phentsize * elf->ehdr->e_phnum, elf->file);
+   if (nread < elf->ehdr->e_phentsize * elf->ehdr->e_phnum) {
+      km_err(2, "Cannot read PHDR %s", path);
+   }
+
+   // Read Section Headers
+   elf->shdr = malloc(elf->ehdr->e_shentsize * elf->ehdr->e_shnum);
+   if (fseek(elf->file, elf->ehdr->e_shoff, SEEK_SET) != 0) {
+      km_err(2, "Cannot seek to SHDR %s", path);
+   }
+   nread = fread(elf->shdr, 1, elf->ehdr->e_shentsize * elf->ehdr->e_shnum, elf->file);
+   if (nread < elf->ehdr->e_shentsize * elf->ehdr->e_shnum) {
+      km_err(2, "Cannot read SHDR %s", path);
+   }
+
+   // Find Symbol table and String Table Indexes
+   elf->symidx = elf->stridx = -1;
+   for (int i = 0; i < elf->ehdr->e_shnum; i++) {
+      if (elf->shdr[i].sh_type == SHT_SYMTAB) {
+         elf->symidx = i;
+      }
+      if (elf->shdr[i].sh_type == SHT_STRTAB && i != elf->ehdr->e_shstrndx) {
+         elf->stridx = i;
+      }
```

Comment:
This is actually available in the `sh_link` field in the `SHT_SYMTAB` section header.